pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,360 @@
+	}
+	if delegate.Mirror == nil {
+		delegate.Mirror = root.Mirror
+	}
+	if delegate.MirrorPercent == nil {
+		delegate.MirrorPercent = root.MirrorPercent
+	}
+	if delegate.MirrorPercentage == nil {
+		delegate.MirrorPercentage = root.MirrorPercentage
+	}
+	if delegate.CorsPolicy == nil {
+		delegate.CorsPolicy = root.CorsPolicy
+	}
+	if delegate.Headers == nil {
+		delegate.Headers = root.Headers
+	}
+	return delegate
+}
+
+// return merged match conditions if not conflicts
+func mergeHTTPMatchRequests(root, delegate []*networking.HTTPMatchRequest) (out []*networking.HTTPMatchRequest, conflict bool) {
+	if len(root) == 0 {
+		return delegate, false
+	}
+
+	if len(delegate) == 0 {
+		return root, false
+	}
+
+	// each HTTPMatchRequest of delegate must find a superset in root.
+	// otherwise it conflicts
+	for _, subMatch := range delegate {
+		foundMatch := false
+		for _, rootMatch := range root {
+			if hasConflict(rootMatch, subMatch) {
+				log.Debugf("HTTPMatchRequests conflict: root %v, delegate %v", rootMatch, subMatch)
+				continue
+			}
+			// merge HTTPMatchRequest
+			out = append(out, mergeHTTPMatchRequest(rootMatch, subMatch))
+			foundMatch = true
+		}
+		if !foundMatch {
+			return nil, true
+		}
+	}
+	if len(out) == 0 {
+		conflict = true
+	}
+	return
+}
+
+func mergeHTTPMatchRequest(root, delegate *networking.HTTPMatchRequest) *networking.HTTPMatchRequest {
+	out := *delegate
+	if out.Name == "" {
+		out.Name = root.Name
+	}
+	if out.Uri == nil {
+		out.Uri = root.Uri
+	}
+	if out.Scheme == nil {
+		out.Scheme = root.Scheme
+	}
+	if out.Method == nil {
+		out.Method = root.Method
+	}
+	if out.Authority == nil {
+		out.Authority = root.Authority
+	}
+	// headers
+	if len(root.Headers) > 0 || len(delegate.Headers) > 0 {
+		out.Headers = make(map[string]*networking.StringMatch)
+	}
+	for k, v := range root.Headers {
+		out.Headers[k] = v
+	}
+	for k, v := range delegate.Headers {
+		out.Headers[k] = v
+	}
+	// withoutheaders
+	if len(root.WithoutHeaders) > 0 || len(delegate.WithoutHeaders) > 0 {
+		out.WithoutHeaders = make(map[string]*networking.StringMatch)
+	}
+	for k, v := range root.WithoutHeaders {
+		out.WithoutHeaders[k] = v
+	}
+	for k, v := range delegate.WithoutHeaders {
+		out.WithoutHeaders[k] = v
+	}
+	// queryparams
+	if len(root.QueryParams) > 0 || len(delegate.QueryParams) > 0 {
+		out.QueryParams = make(map[string]*networking.StringMatch)
+	}
+	for k, v := range root.QueryParams {
+		out.QueryParams[k] = v
+	}
+	for k, v := range delegate.QueryParams {
+		out.QueryParams[k] = v
+	}
+	// sourcelabels
+	if len(root.SourceLabels) > 0 || len(delegate.SourceLabels) > 0 {
+		out.SourceLabels = make(map[string]string)
+	}
+	for k, v := range root.SourceLabels {
+		out.SourceLabels[k] = v
+	}
+	for k, v := range delegate.SourceLabels {
+		out.SourceLabels[k] = v
+	}
+	if len(out.Gateways) == 0 {
+		out.Gateways = root.Gateways
+	}
+	if out.SourceNamespace == "" {
+		out.SourceNamespace = root.SourceNamespace
+	}
```

Comment:
same as above. never use the gateways or source namespaces from delegate. Root could say src namespace foo, while delegate could say src namespace bar. I think we should pick root. Or better yet, we can completely eliminate the source match and say that when doing delegation, source match will not apply.

Reply:
```
	if root.SourceNamespace != "" && leaf.SourceNamespace != "" && root.SourceNamespace != leaf.SourceNamespace {
		return true
	}
```

See L304, if the value is not same, it conflicts

Reply:
same